File: py_tdd/py_tdd/portfolio.py
from dataclasses import dataclass, field
from .money import Money
from .bank import Bank


@dataclass
class Portfolio:
    money: list[Money] = field(default_factory=list)

    def add(self, *moneys: Money) -> None:
        self.money.extend(moneys)

    def evaluate(self, bank: Bank, currency: str) -> Money:
        total = 0.0
        failures: list[Exception] = []
        for m in self.money:
            try:
                total += bank.convert(m, currency).amount
            except Exception as ex:
                failures.append(ex)
        if len(failures) == 0:
            return Money(amount=total, currency=currency)
        failureMessage = ",".join(str(f) for f in failures).replace("'", "")
        raise Exception("Missing exchange rate(s):[" + failureMessage + "]")

File: js/portfolio.js
const Money = require("./money");
class Portfolio {
  constructor() {
    this.moneys = [];
  }

  add(...moneys) {
    this.moneys = this.moneys.concat(moneys);
  }

  convert(money, currency) {
    let eurToUsd = 1.2;
    if (money.currency === currency) {
      return money.amount;
    }
    return money.amount * eurToUsd;
  }

  evaluate(currency) {
    let total = this.moneys.reduce((sum, money) => {
      return sum + this.convert(money, currency);
    }, 0);
    return new Money(total, currency);
  }
}

module.exports = Portfolio;
